vllm/v1/core/scheduler.py

```diff
@@ -278,10 +294,13 @@ def schedule(self) -> "SchedulerOutput":
                     # The request cannot be scheduled.
                     break
 
-                self.waiting.popleft()
+                self.waiting.pop(num_to_skip)
+                if request.use_guided_decoding:
+                    guided_decoding_request_ids[request.request_id] = req_index
+                req_index += 1
                 self.running.append(request)
                 self.scheduled_req_ids.add(request.request_id)
-                if request.status == RequestStatus.WAITING:
+                if RequestStatus.is_waiting(request.status):
```

Comment:
May I ask, if RequestStatus.is_waiting includes WAITING and WAITING_FOR_FSM, why do you still need to update the status from WAITING_FOR_FSM to WAITING in line 238?

Reply:
This is because initially when the FSM is not ready, we still have to add the request into the waiting queue.

The transition from waiting_from_fsm to waiting ensures that the FSM will be ready then.

Reply:
Please correct me if I'm wrong, RequestStatus.is_waiting returns true if the status is either WAITING or WAITING_FOR_FSM. Does it mean even if the grammer isn't ready, the status isn't changed to WAITING, line 318 will still be triggered? So I think you should still use request.status == RequestStatus.WAITING?

Reply:
We will never get here for a request in WAITING_FOR_FSM. We continue the loop in that case about 60-70 lines up from here.

Reply:
> why do you still need to update the status from WAITING_FOR_FSM to WAITING in line 238?

to explicitly mark that the request is no longer blocked on FSM compilation, but waiting to execute

Reply:
We should just change this back. It's not necessary anymore.

```suggestion
                if request.status == RequestStatus.WAITING:
```